sdk/client/src/stream/streaming_client.rs

```diff
@@ -0,0 +1,275 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{stream::websocket_transport::WebsocketTransport, StreamError, StreamResult};
+use diem_json_rpc_types::{
+    stream::{
+        request::{StreamMethodRequest, SubscribeToEventsParams, SubscribeToTransactionsParams},
+        response::StreamJsonRpcResponse,
+    },
+    Id,
+};
+use diem_types::event::EventKey;
+use futures::Stream;
+use std::{
+    collections::HashMap,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+    time::Duration,
+};
+use tokio::{
+    sync::{mpsc, RwLock},
+    time::timeout,
+};
+
+use tokio_tungstenite::tungstenite::protocol::WebSocketConfig;
+use tracing::{debug, warn, trace};
+
+pub(crate) type StreamingClientReceiver = mpsc::Receiver<StreamResult<StreamJsonRpcResponse>>;
+pub(crate) type StreamingClientSender = mpsc::Sender<StreamResult<StreamJsonRpcResponse>>;
+
+struct SubscriptionSender {
+    pub id: Id,
+    pub sender: StreamingClientSender,
+}
+
+impl SubscriptionSender {
+    pub fn new(id: Id, sender: StreamingClientSender) -> Self {
+        Self {
+            id,
+            sender,
+        }
+    }
+}
+
+pub struct SubscriptionStream {
+    id: Id,
+    stream: StreamingClientReceiver,
+}
```

Comment:
You have a `clear_subscription` function defined on the client below, does that need to be called when a Subscription is dropped?

Reply:
no, it'd get handled in the `handle_next_message` (channel closed)